### PR TITLE
Fix duplicate function detection for overloads introduced by aliases

### PR DESCRIPTION
**Pull request by MoonlightSentinel**
The previous implementation compared small parts of the function mangling
    to detect duplicate functions. This approach had several flaws manifesting
    in the referenced issue as well as other false positives for (static)
    members, `alias`ing into different scopes, ... .

    Issue 21505 was caused by the assumption that all functions visited by
    `overloadApply` have the same name. This doesn't hold for overloads
    introduced by an `alias`.

    A reliable mangling-based implementation would need to compare most of
    the mangled name while also omitting severel aspects e.g. the return type
    to detect `int foo()` and `void foo()`.

    Hence the current implementation was replaced by actually comparing the
    `FuncDeclaration`s because it allows for more fine grained checks (and
    also should save some memory allocations).

---

Added [the intial report](https://github.com/dlang/dmd/pull/8429#issuecomment-749815607) as a bugzilla reference

### CODE REVIEW

File: test/fail_compilation/fail2789.d [hunk 1]
```diff
@@ -1,11 +1,5 @@
 /*
 https://issues.dlang.org/show_bug.cgi?id=18385
-
-TEST_OUTPUT:
----
-fail_compilation/fail2789.d(15): Error: function `fail2789.A2789.m()` conflicts with previous declaration at fail_compilation/fail2789.d(10)
-fail_compilation/fail2789.d(25): Error: function `fail2789.A2789.m()` conflicts with previous declaration at fail_compilation/fail2789.d(10)
----
```
Comment (PetarKirov) [inline test/fail_compilation/fail2789.d:8]: I'm confused - how come these functions don't conflict? D doesn't allow overloading on return types alone.

Specifically these two:

```D
class C
{
    int m();
    float m();
}
```
overloading on `const` is ok and the `static` `m` is also fine.

File: test/fail_compilation/fail2789.d [hunk 2]
```diff
@@ -53,7 +47,7 @@ void f4() {}
 void f4() {}    // conflict
 
 void f5() @safe {}
-void f5() @system {}    // conflict
+void f5() @system {}    // no conflict because of attribute based overloading in in extern(D)
```
Comment (PetarKirov) [inline test/fail_compilation/fail2789.d:50]: Yes, `extern(D)` preserves the mangling, but does overload resolution work well, or the user has to resort to hackery (e.g. [externDFunc](https://github.com/dlang/druntime/blob/v2.094.2/src/core/internal/traits.d#L122) to traits)? If the answer is no, I think it's not a bad idea to deprecate such form of likely accidental overloading.

Comment (MoonlightSentinel) [reply test/fail_compilation/fail2789.d:50]: Using `__traits(getOverloads, ...)` should also work IIRC.
But thats out of scope for this PR.

File: test/fail_compilation/fail2789.d (hunk 1 above)
Comment (MoonlightSentinel) [reply test/fail_compilation/fail2789.d:8]: The current implementation is tied to the generated mangling which might include the return type. (AFAIC this was done because functions with identical manglings cause glue/backend errors for ldc/gdc).

Guess this should eventually be reworked to compare the `FuncDeclaration`s s.t. it can detect these invalid overloads.

File: src/dmd/dmangle.d [hunk 3]
```diff
@@ -1202,23 +1202,14 @@ extern (C++) void mangleToBuffer(TemplateInstance ti, OutBuffer* buf)
...
 {
-    auto tf = fd.type.isTypeFunction();
-
     scope Mangler v = new Mangler(&buf);
-
-    MODtoDecoBuffer(&buf, tf.mod);
-    foreach (idx, param; tf.parameterList)
-        param.accept(v);
-    buf.writeByte('Z' - tf.parameterList.varargs);
+    v.mangleExact(fd);
```
Comment (RazvanN7) [inline src/dmd/dmangle.d:1214]: Can you please elaborate why is this change necessary? It seems like this function does not include the return type in the mangled name. This might be reason why the patch allows overloading on return types.

**Comment by RazvanN7** [comment]
An alternative fix would be to disallow inserting an overload in an overload list via alias this if the inserted function might conflict with any of the other overloads. Inserting a conflicting overload can never be useful.

File: src/dmd/dmangle.d (hunk 3 above)
Comment (MoonlightSentinel) [reply src/dmd/dmangle.d:1214]: The initial bug report was just one examples where this function would cause false positives (see e.g. the second example in the test). Fixing all of tthem lead me to re-implementing most of the mangling logic.

But I'll check if this can be achieved without these changes.

File: test/fail_compilation/fail2789.d (hunk 1 above)
Comment (MoonlightSentinel) [reply test/fail_compilation/fail2789.d:8]: Fixed with the new implementation

File: src/dmd/dmangle.d (hunk 3 above)
Comment (MoonlightSentinel) [reply src/dmd/dmangle.d]: Ditched the current approach in favour of actually comparing the `FuncDeclaration`s

**Comment by MoonlightSentinel** [comment]
Switchted to comparing the `FuncDeclaration`s instead of comparing the mangling because the latter doesn't work for several cases aside from the initial bug report.

File: test/fail_compilation/fail2789.d (hunk 2 above)
Comment (RazvanN7) [reply test/fail_compilation/fail2789.d:55]: I think that is kind of a controversial change. I don't really know if D is able to do overload resolution based on attributes. This fails to compile on run.dlang.io:

```d
void func() pure {}
void func() @nogc {}

void main() @nogc
{
    func();
}
```

**Comment by MoonlightSentinel** [comment]
@RazvanN7 You can still call the function using `__traits(getOverloads)`:

```d
extern (C) int puts(scope const char*) @trusted @nogc;

void func() @safe { puts("@safe"); }
void func() @nogc { puts("@nogc"); }

void main() @nogc
{
    __traits(getOverloads, M, "func")[1](); // Writes "@nogc"
}
```

But I can also defer those changes into a latter PR and retain the current behaviour for now.

**Comment by WalterBright** [comment]
With this behavior change, how does it affect the Spec?

**Comment by MoonlightSentinel** [comment]
The spec doesn't mention conflicting functions IIRC. That's why the proposed implementation checks only for functions that cause issues in the backend (function merging for dmd/errors for ldc/gdc).

**Review by RazvanN7 — APPROVED** [review]

**Comment by MoonlightSentinel** [comment]
Rebased & squashed the commits

**Comment by dkorpel** [comment]
I don't know why, but this change introduced a lot of errors in the form of `address of variable x assigned to y with longer lifetime` in my code. I'm working on a reduced test case.

**Comment by dkorpel** [comment]
```D
@safe:

struct I {int* x;}
pure I gcd(I, I) {return I();}

auto foo(I func) {
	if (R n = func)
		R r = n.foo!()(n); // Error: address of variable n assigned to r with longer lifetime`
}

struct R {
@safe:
	I d;
	this(I) {}
	bool opCast() {return true;}

	R foo()(R) out {check();} do {
		gcd(d, d);
		return this;
	}

	bool check() {return true;}
}
```

When compiled with `-dip1000`, it says `Error: address of variable n assigned to r with longer lifetime`.
When reverting this commit, it doesn't happen. I don't know how this PR changed this, but it seems to be a side-effect:  small changes to the snippet make or break the new / old compiler version. I don't even know if this is accepts-invalid or rejects-valid.

**Comment by MoonlightSentinel** [comment]
The DIP1000 implementation is very brittle, so I guess this error is probably depending on the order of semantic analysis.
I guess this PR affects your code because the parameter checks causes an earlier expansion & analysis of the function parameters.
